From d327e5de2fa21c48c0e832fb48a65c90f7cb44b8 Mon Sep 17 00:00:00 2001
From: Vignesh Viswanathan <viswanat@qti.qualcomm.com>
Date: Tue, 23 Jun 2026 15:24:40 +0530
Subject: [PATCH] remoteproc: qcom_q6v5_pas: Fix error masking in adsp_stop()

In adsp_stop function, return value of qcom_pas_shutdown for pas_id
is overwritten by the return value of qcom_pas_shutdown for dtb_pas_id.
This causes errors seen on qcom_pas_shutdown failures for pas_id to be
masked to the caller. This might lead to issues where the memory regions
locked by PAS, as part of qcom_pas_auth_and_reset, are not released for
access by linux and rproc_coredump flow will end up accessing the locked
memory, leading to an access violation.

Fix this by using a separate variable for the dtb_pas_id shutdown call
and only overriding the main return value if the pas_id shutdown succeeded
but dtb_pas_id shutdown failed.

Change-Id: Iad101e5a67d9248f6b5871bd3e8f43824392f252
Signed-off-by: Vignesh Viswanathan <viswanat@qti.qualcomm.com>
---
 drivers/remoteproc/qcom_q6v5_pas.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/drivers/remoteproc/qcom_q6v5_pas.c b/drivers/remoteproc/qcom_q6v5_pas.c
index ed960ab6bdfc..f5cf17c82a08 100644
--- a/drivers/remoteproc/qcom_q6v5_pas.c
+++ b/drivers/remoteproc/qcom_q6v5_pas.c
@@ -401,6 +401,7 @@ static void qcom_pas_handover(struct qcom_q6v5 *q6v5)
 static int adsp_stop(struct rproc *rproc)
 {
 	struct qcom_adsp *adsp = rproc->priv;
+	int dtb_ret;
 	int handover;
 	int ret;
 
@@ -415,10 +416,13 @@ static int adsp_stop(struct rproc *rproc)
 	if (ret)
 		dev_err(adsp->dev, "failed to shutdown: %d\n", ret);
 
+	/* Shutdown DTB processor if present, but preserve main shutdown error */
 	if (adsp->dtb_pas_id) {
-		ret = qcom_pas_shutdown(adsp->dtb_pas_id);
-		if (ret)
-			dev_err(adsp->dev, "failed to shutdown dtb: %d\n", ret);
+		dtb_ret = qcom_pas_shutdown(adsp->dtb_pas_id);
+		if (dtb_ret)
+			dev_err(adsp->dev, "failed to shutdown dtb: %d\n", dtb_ret);
+		if (!ret && dtb_ret)
+			ret = dtb_ret;
 	}
 
 	handover = qcom_q6v5_unprepare(&adsp->q6v5);
-- 
2.34.1

